### PR TITLE
changefeedccl: Introduce cdcevent package.

### PR DESCRIPTION
**Pull request by miretskiy**
Event processing in CDC is a fairly involved procedure, consisting
of multiple stages.  A low level KeyValue event is received from
rangefeeds.  Then this low level event needs to be decoded into
encoded datums, along with type information describing those datums,
representing the "row" in the table.
Thereafter those datums must be further encoded into appropriate format.

Historically, this pipeline operated on structures representing
fairly low level objects -- such as catalog.Descriptor, EncDatumRow, etc.

This is problematic for multiple reasons:
  * A tight coupling to the underlying table/family descriptors makes it
    hard to introduce layers into pipeline that may not be directly tied
    to those descriptors -- such as projections and predicates.
  * Exposing low level descriptor information meant that every encoder
    implementation had to have repeated, and very subtle and error prone code,
    to process columns in the correct order -- filtering unwanted
    families, processing primary index column in the correct order, etc.
  * The catalog.Descriptor interfaces are fairly large interfaces,
    understandably so.  However, CDC does not need access to the
    entirety of those interfaces -- mostly, CDC needs to access to very few
    methods.  It is desirable therefore to pass in the smallest
    interface necessary to make CDC work.

This is the purpose of `cdcevent` package.  It introduces several
abstractions to facilitate conversion from the low level KeyValue event
into `cdcevent.EventRow` event.  The `cdcevent.EventDecoder` interface
defines conversion from low level event into `cdcevent.EventRow`.  Event
row has associated event descriptor describing the set of columns needed
to perform CDC activities.  The rest of the CDC processing operates
on these higher level interfaces instead of having direct depency on
low level catalog interfaces.

Release Notes: None

### CODE REVIEW

**Review by HonoreDB — COMMENTED** [review]
Thank you for taking this on! Couple of minor comments. Also, is there an easy way to verify that introducing this extra layer of abstraction doesn't result in more allocs on the hot path?

> Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, 20 of 20 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81249#-:-N1zLZaRApzoRor6APnC:b-2tdq69)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 47 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81249#-N1zHaKp3mMg4F8T27E8:-N1zHaKp3mMg4F8T27E9:b3ur1z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/067c906ac0a45f677a958ba4aa61878182923d5d/pkg/ccl/changefeedccl/cdcevent/event.go#L47)):*
> ```Go
> 	FamilyName() string
> 	// NumFamilies returns number of families in the table.
> 	NumFamilies() int
> ```

nit: could this be `TableHasOtherFamilies() bool`? Both because the EventSource always only has one family, and the only information we actually need is whether there are others.
___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 380 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/81249#-N1zK15z2biJuJ4CcXMi:-N1zK15z2biJuJ4CcXMj:bjnxhlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/067c906ac0a45f677a958ba4aa61878182923d5d/pkg/ccl/changefeedccl/cdcevent/event.go#L380)):*
> ```Go
> 		return EventRow{}, err
> 	}
> 	// Copy datums since row fetcher reuses alloc.
> ```

nit: can remove this comment

File: pkg/ccl/changefeedccl/cdcevent/event.go [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+	"github.com/cockroachdb/cockroach/pkg/sql/row"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/cache"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
+	"github.com/cockroachdb/errors"
+)
+
+// EventSource identifies the source of the event.
```
Comment (ajwerner) [inline pkg/ccl/changefeedccl/cdcevent/event.go:32]: I'm not in love with this as an interface. Effectively all of the methods are constants. The ones that aren't change very rarely. How would you feel about making this a struct? Like cdcevent.Metadata or Info or something? I also find the name of it to be pretty bad. It's not a source with these methods. 

I don't think you'll have to change much of anything to export the eventDescriptor struct as it is. I think it'd even be fine to embed that newly exported struct and leave all of its fields private as they currently are. You'd need to change almost nothing.

File: pkg/ccl/changefeedccl/cdcevent/event.go [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+	Version() descpb.DescriptorVersion
+	// FamilyID returns familyID for the event.
+	FamilyID() descpb.FamilyID
+	// FamilyName returns family name.
+	FamilyName() string
+	// NumFamilies returns number of families in the table.
+	NumFamilies() int
+}
+
+// EventDecoder is an interface for decoding KVs into cdc event row.
```
Comment (ajwerner) [inline pkg/ccl/changefeedccl/cdcevent/event.go:50]: Nit: just call it Decoder

File: pkg/ccl/changefeedccl/cdcevent/event.go [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+type EventDecoder interface {
+	// DecodeKV decodes specified key value to EventRow.
+	DecodeKV(ctx context.Context, kv roachpb.KeyValue, schemaTS hlc.Timestamp) (EventRow, error)
+}
+
+// EventRow holds a row corresponding to an event.
+type EventRow struct {
```
Comment (ajwerner) [inline pkg/ccl/changefeedccl/cdcevent/event.go:57]: Nit: just call it Row

File: pkg/ccl/changefeedccl/cdcevent/event.go [hunk 4]
```diff
@@ -0,0 +1,491 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package cdcevent
```
Comment (ajwerner) [inline pkg/ccl/changefeedccl/cdcevent/event.go:9]: Give us a doc.go and a comment explaining how this package fits into the puzzle

**Review by ajwerner — COMMENTED** [review]
Subway first pass.

**Comment by miretskiy** [comment]
> nit: could this be `TableHasOtherFamilies() bool`? Both because the EventSource always only has one family, and the only information we actually need is whether there are others.

Done.

> nit: can remove this comment

Done.

File: pkg/ccl/changefeedccl/cdcevent/event.go (hunk 2 above)
Comment (miretskiy) [reply pkg/ccl/changefeedccl/cdcevent/event.go:50]: Done.

File: pkg/ccl/changefeedccl/cdcevent/event.go (hunk 3 above)
Comment (miretskiy) [reply pkg/ccl/changefeedccl/cdcevent/event.go:57]: Done.

**Comment by miretskiy** [comment]
> Also, is there an easy way to verify that introducing this extra layer of abstraction doesn't result in more allocs on the hot path?

I don't think there is an easy way; I've tried not to do anything too silly.  However, this should not prevent from merging (eventually) this pr.. We can improve things if needed.

File: pkg/ccl/changefeedccl/cdcevent/event.go (hunk 1 above)
Comment (miretskiy) [reply pkg/ccl/changefeedccl/cdcevent/event.go:32]: > I'm not in love with this as an interface. Effectively all of the methods are constants. The ones that aren't change very rarely. How would you feel about making this a struct? Like cdcevent.Metadata or Info or something? I also find the name of it to be pretty bad. It's not a source with these methods.
> 
> I don't think you'll have to change much of anything to export the eventDescriptor struct as it is. I think it'd even be fine to embed that newly exported struct and leave all of its fields private as they currently are. You'd need to change almost nothing.

Name -- I don't mind -- it was the source for the EventRow (or Row) as it's being renamed.
It's alright to change things. Changing it to struct makes sense.  
I think the reason for it being an interface was an implementation, I think in cloud storage sink test, that mucked with version... But... that's even easier to accomplish with a struct containing public (constant) members.

File: pkg/ccl/changefeedccl/cdcevent/event.go (hunk 4 above)
Comment (miretskiy) [reply pkg/ccl/changefeedccl/cdcevent/event.go:9]: Done.

**Comment by miretskiy** [comment]
@HonoreDB @ajwerner all comments addressed.

**Review by HonoreDB — APPROVED** [review]
> Reviewed 11 of 15 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81249)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 183 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/81249#-N2DSEjy3AGlIp9WBsF2:-N2DSEjy3AGlIp9WBsF3:bz5akah) ([raw file](https://github.com/cockroachdb/cockroach/blob/feb8dfaa894c0efe5ee110048949380686560d51/pkg/ccl/changefeedccl/cdcevent/event.go#L183)):*
> ```Go
> 	desc catalog.TableDescriptor, family *descpb.ColumnFamilyDescriptor, includeVirtualColumns bool,
> ) (*eventDescriptor, error) {
> 	var inFamily catalog.TableColSet
> ```

golf: `inFamily := catalog.MakeTableColSet(family.ColumnIDs...)`

**Comment by miretskiy** [comment]
> inFamily := catalog.MakeTableColSet(family.ColumnIDs...)

Nice; thanks

**Review by ajwerner — APPROVED** [review]
Cursory second pass :lgtm: 

> Reviewed 1 of 15 files at r6, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81249#-:-N2E03tH62JnfViO1GOO:bd0v8z3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdcevent/event.go` line 244 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/81249#-N2E-VkeD988YWiL6U1A:-N2E-VkeD988YWiL6U1B:b8zrt2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8c6f1fa31c322711b2350dd48750e0491d2984a/pkg/ccl/changefeedccl/cdcevent/event.go#L244)):*
> ```Go
> 
> // DebugString returns event descriptor debug information.
> func (d *eventDescriptor) DebugString() string {
> ```

if you felt fancy, you could implement `SafeFormat` or whatever the method name is so that this object and even the row itself can be logged. Might come in handy 



_Code quote:_
```Go
// DebugString returns event descriptor debug information.
func (d *eventDescriptor) DebugString() string {
	return fmt.Sprintf("eventDescriptor{table: %q(%d) family: %q(%d) pkCols=%v valCols=%v",
		d.TableName, d.TableID, d.FamilyName, d.FamilyID, d.keyCols, d.familyCols)
}
```

**Comment by miretskiy** [comment]
> if you felt fancy, you could implement `SafeFormat`

added

File: pkg/ccl/changefeedccl/encoder_csv.go [hunk 5]
```diff
@@ -12,33 +12,20 @@ import (
 	"bytes"
 	"context"
 
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdcevent"
 	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
```
Comment (shermanCRL) [inline pkg/ccl/changefeedccl/encoder_csv.go:17]: Interesting bit after our conversation about the abstraction layer: what would it take to have zero `/sql` dependency in the encoders?

**Comment by miretskiy** [comment]
We still need access  to e.g. descpb packages; (access to tableID types,
etc).

File: pkg/ccl/changefeedccl/encoder_csv.go (hunk 5 above)
Comment (ajwerner) [reply pkg/ccl/changefeedccl/encoder_csv.go:17]: `tree` includes datum definitions. All of the row data is given to the encoders in terms of datums. Breaking that one and `sql/types` isn't worth it.

**Pull request merged**